File: algo/16. 3Sum Closest.py
class Solution:
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        
        if len(nums) < 3:
            return []
        
        nums.sort()
        
        minDist = None
        closeSum = 0
        
        for i in range(len(nums)-2):
            currNum = nums[i]
            left = i+1
            right = len(nums)-1
            while (left!=right):
                currSum = currNum + nums[left] + nums[right]
                currDist = abs(currSum - target)
                if minDist is None:
                    minDist = currDist
                    closeSum = currSum
                elif currDist < minDist:
                    minDist = currDist
                    closeSum = currSum
                
                if currSum == target:
                    return currSum
                elif currSum < target:
                    left += 1
                elif currSum > target:
                    right -= 1
        return closeSum
        
        
        

File: round2/16. 3Sum Closest.cs
public class Solution {
    public int ThreeSumClosest(int[] nums, int target) {
        Array.Sort(nums);
        int? minDiff = null;
        int currSum, res=0;
        for (int i=0; i<nums.Length-2; i++) {
            int left = i+1, right = nums.Length-1;
            while (left != right) {
                currSum = nums[i]+nums[left]+nums[right];
                if (minDiff == null) {
                    minDiff = Math.Abs(target - currSum);
                    res = currSum;
                } else if (Math.Abs(target - currSum) < minDiff) {
                    minDiff = Math.Abs(target - currSum);
                    res = currSum;
                }
                if (currSum == target) {
                    return target;
                } else if (currSum < target) {
                    left ++;
                } else {
                    right --;
                }
            }
        }
        return res;
    }
}